File: backend/app/schemas/review.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional, Union
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class ReviewIssueSchema(BaseModel):
    issue_id: Optional[str] = None
    severity: str
    line_number: int
    description: str
    rule_type: str
    ai_explanation: Optional[str] = None
    suggestion: Optional[str] = None
    source: Optional[str] = None  # "Static", "AI", or "Static + AI"
    confidence: Optional[int] = None
    file: Optional[str] = None


class ReviewBase(BaseModel):
    title: str = Field(..., max_length=255)
    language_id: Optional[Union[UUID, str]] = None


class ReviewCreateRequest(ReviewBase):
    source_code: Optional[str] = None
    repo_url: Optional[str] = Field(None, max_length=1024)
    file_name: Optional[str] = Field(None, max_length=255)
    file_size: Optional[int] = None

    @model_validator(mode="after")
    def check_source_or_repo(self) -> ReviewCreateRequest:
        if not self.source_code and not self.repo_url:
            raise ValueError("Either source_code or repo_url must be provided.")
        return self


class LanguageSchema(BaseModel):
    id: UUID
    name: str
    model_config = ConfigDict(from_attributes=True)

class ReviewResponse(ReviewBase):
    id: UUID
    user_id: UUID
    source_code: str
    improved_code: Optional[str] = None
    issues: List[Dict[str, Any]] = Field(default_factory=list)
    quality_score: Optional[int] = None
    language: Optional[LanguageSchema] = None
    review_metadata: Dict[str, Any] = Field(default_factory=dict, serialization_alias="metadata")
    status: str
    file_name: Optional[str] = None
    file_size: Optional[int] = None
    repo_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ReviewListResponse(BaseModel):
    items: List[ReviewResponse]
    total: int
    page: int
    size: int

File: frontend/src/types/review.ts
export type IssueSeverity = 'Critical' | 'High' | 'Medium' | 'Low';

export type IssueType = 'Bugs' | 'Smells' | 'Security' | 'SOLID' | 'Complexity' | 'Optimization' | 'Maintainability' | 'Best Practices' | 'Naming';

export interface Issue {
  issue_id?: string;
  severity: IssueSeverity;
  line_number: number;
  description: string;
  rule_type: string;
  ai_explanation?: string;
  file?: string;
  suggestion?: string;
  source?: 'Static' | 'AI' | 'Static + AI';
  confidence?: number;
}

export interface LanguageDetectionResult {
  selected_language: string;
  detected_language: string;
  confidence: number;
  is_match: boolean;
  evidence: string[];
  final_language?: string;
  language_switched?: boolean;
}

export interface ReviewMetadata {
  quality_score?: number;
  maintainability_grade?: string;
  tech_debt?: string;
  cyclomatic_complexity?: number;
  est_refactoring_time?: string;
  security_score?: number;
  performance_score?: number;
  ai_summary?: string;
  issue_counts?: {
    critical: number;
    high: number;
    medium: number;
    low: number;
  };
  repo_insights?: {
    file_count: number;
    largest_files: Array<{name: string; size: number}>;
    language_distribution: Record<string, string>;
    repo_health_score: number;
  };
  scoring_engine?: {
    version: string;
    overall_quality: number;
    security_score: number;
    performance_score: number;
    maintainability_score: number;
    maintainability_grade: string;
    technical_debt_score: number;
  };
  language_detection?: LanguageDetectionResult;
  ai_status?: 'available' | 'unavailable';
  ai_unavailable_reason?: 'rate_limit' | 'provider_error' | string;
}

export interface Review {
  id: string;
  user_id: string;
  language_id: string;
  language?: { id: string; name: string };
  title: string;
  source_code: string;
  improved_code?: string;
  issues: Issue[];
  quality_score?: number;
  metadata: ReviewMetadata;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  file_name?: string;
  file_size?: number;
  repo_url?: string;
  created_at: string;
  updated_at: string;
}

export interface ReviewCreatePayload {
  title: string;
  language_id?: string;
  source_code?: string;
  repo_url?: string;
}

export interface ReviewFilter {
  status?: string;
  language?: string;
  skip?: number;
  limit?: number;
  maxScore?: number;
}

export interface ReviewSummary {
  id: string;
  repositoryUrl: string;
  branch: string;
  status: string;
  overallScore: number;
  issuesFound: number;
  criticalIssues: number;
  createdAt: string;
}

